engines/sci/graphics/video32.cpp

```diff
@@ -0,0 +1,393 @@
+
+		_decoder->start();
+	}
+
+	EventFlags stopFlag = kEventFlagNone;
+	while (!g_engine->shouldQuit()) {
+		if (_decoder->endOfVideo()) {
+			stopFlag = kEventFlagEnd;
+			break;
+		}
+
+		g_sci->getEngineState()->speedThrottler(_decoder->getTimeToNextFrame());
+		g_sci->getEngineState()->_throttleTrigger = true;
+		if (_decoder->needsUpdate()) {
+			renderFrame();
+		}
+
+		const int currentFrameNo = _decoder->getCurFrame();
+
+		if (
+			_yieldInterval > 0 &&
+			currentFrameNo != _lastYieldedFrameNo &&
+			(currentFrameNo % _yieldInterval) == 0
+		) {
+			_lastYieldedFrameNo = currentFrameNo;
+			stopFlag = kEventFlagYieldToVM;
+			break;
+		}
+
+		const SciEvent event = _eventMan->getSciEvent(SCI_EVENT_MOUSE_PRESS | SCI_EVENT_KEYBOARD);
+
+		if ((flags & kEventFlagMouseDown) && event.type == SCI_EVENT_MOUSE_PRESS) {
+			stopFlag = kEventFlagMouseDown;
+			break;
+		}
+
+		if ((flags & kEventFlagEscapeKey) && event.type == SCI_EVENT_KEYBOARD && event.character == SCI_KEY_ESC) {
+			stopFlag = kEventFlagEscapeKey;
+			break;
+		}
+
+		// TODO: Hot rectangles
+		if ((flags & kEventFlagHotRectangle) /* && event.type == SCI_EVENT_HOT_RECTANGLE */) {
+			warning("Hot rectangles not implemented in VMD player");
+			stopFlag = kEventFlagHotRectangle;
+			break;
+		}
+	}
+
+	_decoder->pauseVideo(true);
+	return stopFlag;
+}
+
+#pragma mark -
+#pragma mark VMDPlayer - Rendering
+
+void VMDPlayer::renderFrame() const {
+	// This writes directly to the CelObjMem we already created,
+	// so no need to take its return value
+	_decoder->decodeNextFrame();
+
+	// NOTE: Normally this would write a hunk palette at the end of the
+	// video bitmap that CelObjMem would read out and submit, but instead
+	// we are just submitting it directly here because the decoder exposes
+	// this information a little bit differently than the one in SSCI
+	const bool dirtyPalette = _decoder->hasDirtyPalette();
+	if (dirtyPalette) {
+		Palette palette;
+		palette.timestamp = g_sci->getTickCount();
+		if (_blackPalette) {
+			for (uint16 i = _startColor; i <= _endColor; ++i) {
+				palette.colors[i].r = palette.colors[i].g = palette.colors[i].b = 0;
+				palette.colors[i].used = true;
+			}
+		} else {
+			fillPalette(palette);
+		}
+
+		g_sci->_gfxPalette32->submit(palette);
+		g_sci->_gfxFrameout->updateScreenItem(*_screenItem);
+		g_sci->_gfxFrameout->frameOut(true);
+
+		if (_blackPalette) {
+			fillPalette(palette);
+			g_sci->_gfxPalette32->submit(palette);
+			g_sci->_gfxPalette32->updateForFrame();
+			g_sci->_gfxPalette32->updateHardware();
+		}
+	} else {
+		g_sci->_gfxFrameout->updateScreenItem(*_screenItem);
+		g_sci->getSciDebugger()->onFrame();
+		g_sci->_gfxFrameout->frameOut(true);
+		g_sci->_gfxFrameout->throttle();
+	}
+}
+
+void VMDPlayer::fillPalette(Palette &palette) const {
+	const byte *vmdPalette = _decoder->getPalette() + _startColor * 3;
+	for (uint16 i = _startColor; i <= _endColor; ++i) {
+		int16 r = *vmdPalette++;
```

Comment:
Hmm. I don’t *think* it should matter, the largest calculated value should always be well below the maximum allowed value in either case.